<a href="https://colab.research.google.com/github/meghna2312/Diabetesprediction/blob/master/AIProjectNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Project by:
Meghna Ashok
16070122034 

Aim: **To predict if a person has diabetes(1) or no diabetes(0)**

The pima-indians-diabetes data set comes from the Pima people, who are a group of Native Americans living in an area consisting of what is now central and southern Arizona. They have the highest reported prevalence of diabetes of any population in the world, and have contributed to numerous scientific gains through their willingness to participate in research.




In [ ]:
from google.colab import files
uploaded = files.upload() 

Saving diabetes.csv to diabetes.csv


In [ ]:
from keras.models import Sequential
from keras.layers import Dense
import pandas as pd
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')

In [ ]:
df = pd.read_csv('diabetes.csv')
df.head(10)

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1
5            5      116             74  ...                     0.201   30        0
6            3       78             50  ...                     0.248   26        1
7           10      115              0  ...                     0.134   29        0
8            2      197             70  ...                     0.158   53        1
9            8      125             96  ...                     0.232   54        1

[10 rows x 9 columns]

In [ ]:
df.shape

(768, 9)

In [ ]:
df.drop_duplicates(inplace = True)

In [ ]:
dataset = df.values

In [ ]:
X = dataset[:,0:8] 
# last column i.e target is y
y = dataset[:,8]

In [ ]:
#data processing so that all features are between 0 and 1
from sklearn import preprocessing
min_max_scaler = preprocessing.MinMaxScaler()
X_scale = min_max_scaler.fit_transform(X)
X_scale

array([[0.35294118, 0.74371859, 0.59016393, ..., 0.50074516, 0.23441503,
        0.48333333],
       [0.05882353, 0.42713568, 0.54098361, ..., 0.39642325, 0.11656704,
        0.16666667],
       [0.47058824, 0.91959799, 0.52459016, ..., 0.34724292, 0.25362938,
        0.18333333],
       ...,
       [0.29411765, 0.6080402 , 0.59016393, ..., 0.390462  , 0.07130658,
        0.15      ],
       [0.05882353, 0.63316583, 0.49180328, ..., 0.4485842 , 0.11571307,
        0.43333333],
       [0.05882353, 0.46733668, 0.57377049, ..., 0.45305514, 0.10119556,
        0.03333333]])

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_scale, y, test_size=0.2, random_state = 4)

In [ ]:
from tensorflow.python.keras.layers import Dense
from tensorflow.python.keras import Sequential

model = Sequential([
    Dense(12, activation='relu', input_shape=( 8 ,)),
    Dense(15, activation='relu'),
    Dense(1, activation='sigmoid')
])

In [ ]:
model.compile(optimizer='sgd', #Stochastic gradient descent optimizer.
              loss='binary_crossentropy', #Used for binary classification
              metrics=['accuracy'])

In [ ]:
#Train model, split 20% validation data
hist = model.fit(X_train, y_train,cbatch_size=57, epochs=1000, validation_split=0.2)

Epoch 1/1000
16/16 [==============================] - 0s 13ms/step - loss: 0.7108 - accuracy: 0.3585 - val_loss: 0.7034 - val_accuracy: 0.3740
Epoch 2/1000
16/16 [==============================] - 0s 3ms/step - loss: 0.7006 - accuracy: 0.3870 - val_loss: 0.6937 - val_accuracy: 0.5122
Epoch 3/1000
16/16 [==============================] - 0s 3ms/step - loss: 0.6934 - accuracy: 0.5540 - val_loss: 0.6864 - val_accuracy: 0.5935
Epoch 4/1000
16/16 [==============================] - 0s 3ms/step - loss: 0.6868 - accuracy: 0.6191 - val_loss: 0.6807 - val_accuracy: 0.6504
Epoch 5/1000
16/16 [==============================] - 0s 3ms/step - loss: 0.6798 - accuracy: 0.6477 - val_loss: 0.6762 - val_accuracy: 0.6667
Epoch 6/1000
16/16 [==============================] - 0s 3ms/step - loss: 0.6768 - accuracy: 0.6538 - val_loss: 0.6718 - val_accuracy: 0.6423
Epoch 7/1000
16/16 [==============================] - 0s 3ms/step - loss: 0.6714 - accuracy: 0.6497 - val_loss: 0.6676 - val_accuracy: 0.6585
Epoch

In [ ]:
prediction = model.predict(X_test)
prediction  = [1 if y>=0.5 else 0 for y in prediction] #Threshold
print(prediction)
print(y_test)

[0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1]
[0. 0. 0. 0. 1. 1. 1. 1. 1. 0. 0. 0. 1. 0. 0. 1. 1. 1. 0. 0. 0. 0. 0. 1.
 1. 0. 0. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1.
 0. 0. 1. 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0. 1. 0. 1.
 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 0. 1. 0. 0. 0. 1. 0. 1. 0. 0. 0. 0. 0. 0.
 0. 1. 1. 0. 0. 0. 1. 1. 1. 0. 0. 1. 1. 0. 0. 1. 1. 0. 0. 0. 0. 0. 1. 1.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 1. 0. 1. 0. 1. 0. 0. 0. 1. 0. 0. 0. 0. 0.
 1. 1. 1. 0. 0. 0. 0. 0. 1. 1.]


In [ ]:
#Model Evaluation on training set
from sklearn.metrics import classification_report,confusion_matrix, accuracy_score
pred = model.predict(X_train)
pred  = [1 if y>=0.5 else 0 for y in pred] #Threshold
print(classification_report(y_train ,pred ))
print('Confusion Matrix: \n',confusion_matrix(y_train,pred))
print()
print('Accuracy: ', accuracy_score(y_train,pred))
print()


              precision    recall  f1-score   support

         0.0       0.80      0.88      0.84       398
         1.0       0.74      0.59      0.66       216

    accuracy                           0.78       614
   macro avg       0.77      0.74      0.75       614
weighted avg       0.78      0.78      0.78       614

Confusion Matrix: 
 [[352  46]
 [ 88 128]]

Accuracy:  0.7817589576547231



Overall accuracy is approximately 78%

In [ ]:
#Cross validation with test set
from sklearn.metrics import classification_report,confusion_matrix, accuracy_score
pred = model.predict(X_test)
pred  = [1 if y>=0.5 else 0 for y in pred] #Threshold
print(classification_report(y_test ,pred ))
print('Confusion Matrix: \n',confusion_matrix(y_test,pred))
print()
print('Accuracy: ', accuracy_score(y_test,pred))
print()

              precision    recall  f1-score   support

         0.0       0.83      0.83      0.83       102
         1.0       0.67      0.65      0.66        52

    accuracy                           0.77       154
   macro avg       0.75      0.74      0.74       154
weighted avg       0.77      0.77      0.77       154

Confusion Matrix: 
 [[85 17]
 [18 34]]

Accuracy:  0.7727272727272727



Thus, with cross validation with test set as well,the model has approximately 77% accuracy. The accuracy is on average about 1% less than on training set due to some bias. By tuning hyperparameters of model, bias can be reduced, and even accuracy can be potentially improved.
